### PR TITLE
Add scoring mode to MistralCausalLM

### PR DESCRIPTION
**Pull request by RyanMullins**
Adds the `.score()` function introduced with Gemma (https://github.com/keras-team/keras-nlp/pull/1448) to the `MistralCausalLM` model class. As with Gemma, this function supports a variety of interpretability use cases with Mistral by providing an API by which generated sequences can be scored (logits or loss) with gradient tracking on. Use cases include [salience maps](https://pair-code.github.io/lit/documentation/components.html#sequence-salience), [patching](https://arxiv.org/abs/2401.06102), and [training data attribution](https://arxiv.org/abs/2002.08484). 

This is a direct port of the implementation and tests from Gemma, so hopefully that helps ease the review process.

### CODE REVIEW

File: keras_nlp/models/mistral/mistral_causal_lm.py [hunk 1]
```diff
@@ -215,6 +215,132 @@ def next(prompt, cache, index):
...
+                uses `keras.ops.ones()` to create a tensor of the appropriate
+                shape.
+            scoring_mode: The type of scores to return, either "logits" or
+                "loss", both will be per input token.
+            layer_intercept_fn: An optional function for augmenting activations
+                with additional computation, for example, as part of
+                interpretability research. This function will be passed the
+                activations as its first parameter and a numeric index
+                associated with that backbone layer. _This index _is not_ an
+                index into `self.backbone.layers`_. The index -1 accompanies the
```
Comment (mattdangerw) [inline keras_nlp/models/mistral/mistral_causal_lm.py:246]: Extraneous underscore after layers.

**Review by mattdangerw — APPROVED** [review]
LGTM thanks! Just see one tiny nit.

File: keras_nlp/models/mistral/mistral_causal_lm.py (hunk 1 above)
Comment (RyanMullins) [reply keras_nlp/models/mistral/mistral_causal_lm.py:246]: Done.
